<a href="https://colab.research.google.com/github/ortzofi/US-Election-Analysis-With-Census-Data/blob/main/Data_Preparation_Public_Version.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import requests
import numpy as np
import pandas as pd

from google.colab import drive

pd.options.display.max_columns = None

We retrieve the 2016-2020 American Community Survey 5-year data using the US Census API. The full list of variables is available in the [documentation](https://api.census.gov/data/2020/acs/acs5/profile/groups.html). The 'PE' suffix in a variable code stands for percentage and 'E' stands for estimate.

Every state in the US has a two-digit [FIPS code](https://en.wikipedia.org/wiki/Federal_Information_Processing_Standard_state_code) that we will use to get data for it. For example, the code of California is 06 and the code of New York is 36. 


In [ ]:
def get_census_data(key, vars, states):
  """
  Parameters:
  key (str): Census API key
  vars (dict): Dictionary in which each pair is of the form (ACS variable code, Name)
  states (list): List of FIPS state codes

  Returns:
  A dataframe
  """
  get_vars = ",".join(vars.keys())
  get_states = ",".join(states)

  url = "https://api.census.gov/data/2020/acs/acs5/profile?get=NAME," + get_vars + "&for=tract:*&in=state:" + get_states + "&in=county:*&key=" + API_KEY
  response = requests.request("GET", url)

  new_cols = list(vars.values())
  new_cols.insert(0, "Name")
  new_cols.extend(["State", "County", "Tract"])

  return pd.DataFrame(response.json()[1:], columns=new_cols)

In [ ]:
API_KEY = "Your API Key"

get_vars_dict = {
    "DP05_0001E": "Total Population",
    "DP05_0002E": "Men",
    "DP05_0003E": "Women",
    "DP05_0018E": "Median Age",
    "DP05_0037PE": "White",
    "DP05_0038PE": "Black",
    "DP05_0039PE": "Native American",
    "DP05_0044PE": "Asian",
    "DP05_0071PE": "Latin American",
    "DP02_0097E": "Not Citizen",
    "DP02_0030E": "Divorced Men",
    "DP02_0036E": "Divorced Women",
    "DP02_0062PE": "High School Graduate",
    "DP02_0065PE": "Bachelor's Degree",
    "DP02_0066PE": "Graduate Degree",
    "DP03_0062E": "Median Household Income",
    "DP03_0088E": "Per Capita Income",
    "DP03_0061PE": "Household Income 200k+",
    "DP03_0005PE": "Unemployment",
    "DP03_0027PE": "Occupation: Management, Business, Science, Art",
    "DP03_0028PE": "Occupation: Service",
    "DP03_0029PE": "Occupation: Sales and Office",
    "DP03_0030PE": "Occupation: Natural Resources, Construction, Maintenance",
    "DP03_0031PE": "Occupation: Production, Transportation, Material Moving",
    "DP03_0047PE": "Private Sector",
    "DP03_0048PE": "Public Sector",
    "DP03_0049PE": "Self-Employed",
    "DP03_0099PE": "No Health Insurance Coverage",
    "DP03_0128PE": "Poverty",
    "DP03_0020PE": "Commuting: Driving Alone",
    "DP03_0021PE": "Commuting: Carpool",
    "DP03_0022PE": "Commuting: Public Transportation",
    "DP03_0023PE": "Commuting: Walking to Work",
    "DP03_0024E": "Working from Home",
    "DP03_0025E": "Mean Commute Time (Minutes)"
}

states = ["36"]

In [ ]:
ny_df = get_census_data(API_KEY, get_vars_dict, states)
ny_df

Name Total Population   Men  \
0          Census Tract 406, Cayuga County, New York             3400  1685   
1          Census Tract 407, Cayuga County, New York             3633  1750   
2          Census Tract 408, Cayuga County, New York             4668  2712   
3          Census Tract 409, Cayuga County, New York             3683  1930   
4       Census Tract 410.01, Cayuga County, New York             3088  1679   
...                                              ...              ...   ...   
5406  Census Tract 1907.14, Suffolk County, New York             1822   831   
5407  Census Tract 1908.01, Suffolk County, New York             1399   696   
5408  Census Tract 1908.02, Suffolk County, New York             1902   935   
5409  Census Tract 2009.01, Suffolk County, New York              920   496   
5410  Census Tract 2009.03, Suffolk County, New York             1353   756   

     Women Median Age White Black Native American Asian Latin American  \
0     1715       48.0  94.5   3.8             0.0   0.2            1.6   
1     1883       48.7  97.8   0.0             0.0   0.0            0.8   
2     1956       42.2  84.1  11.3             0.0   0.8            2.6   
3     1753       47.0  98.2   0.0             0.1   0.4            2.6   
4     1409       44.8  94.0   0.0             0.0   0.7            5.0   
...    ...        ...   ...   ...             ...   ...            ...   
5406   991       48.8  80.5  14.5             0.0   2.5           18.0   
5407   703       56.6  92.3   2.0             0.0   5.1            4.0   
5408   967       51.0  88.9  10.0             0.8   0.0            7.4   
5409   424       61.1  84.8   2.4             0.0   2.0           17.5   
5410   597       48.3  85.8   5.5             0.0   1.0           13.4   

     Not Citizen Divorced Men Divorced Women High School Graduate  \
0             33           91            186                 28.4   
1             35          112             74                 18.7   
2             46          239            194                 30.4   
3             31          243            158                 43.9   
4             73           82            133                 33.7   
...          ...          ...            ...                  ...   
5406         182           94            134                 33.1   
5407         103           22             32                 10.0   
5408          92           89            112                 19.8   
5409          81           39             28                 14.6   
5410         119           32            131                 21.0   

     Bachelor's Degree Graduate Degree Median Household Income  \
0                 20.6            14.0                   84330   
1                 19.7            19.6                   93493   
2                 11.1             5.1                   64811   
3                 12.9             6.8                   66711   
4                 15.9            11.9                   73182   
...                ...             ...                     ...   
5406              27.5            20.4                   91424   
5407              23.9            53.1                  130714   
5408              42.5            20.0                  124417   
5409              33.5            33.0                   86705   
5410              33.7            18.6                  127250   

     Per Capita Income Household Income 200k+ Unemployment  \
0                38098                    5.5          2.5   
1                48435                   11.5          2.0   
2                25305                    2.6          1.1   
3                32477                    6.6          2.2   
4                35962                    6.3          2.1   
...                ...                    ...          ...   
5406             53240                   25.0          2.9   
5407            227989                   39.0          0.6   
5408             54947        

In [ ]:
ny_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5411 entries, 0 to 5410
Data columns (total 39 columns):
 #   Column                                                    Non-Null Count  Dtype 
---  ------                                                    --------------  ----- 
 0   Name                                                      5411 non-null   object
 1   Total Population                                          5411 non-null   object
 2   Men                                                       5411 non-null   object
 3   Women                                                     5411 non-null   object
 4   Median Age                                                5411 non-null   object
 5   White                                                     5411 non-null   object
 6   Black                                                     5411 non-null   object
 7   Native American                                           5411 non-null   object
 8   Asian                       

We clean the data in a few stages:

1. As all the columns are strings, we have to convert them to integers and floats.

2. Merge the columns of divorced men and women and calculate the percentage of divorcees.

3. Create a column with a combined 11-digit FIPS code representing the census tract of each row. It is done by concatenating the codes of the two-digit state code, the three-digit county code, and the six-digit tract code.

4. Finally, we will sort the dataframe according to the 11-digit FIPS code.

In [ ]:
def convert_columns_dtypes(df):
  # Converting needed columns from strings to integers and floats
  integer_cols = ["Total Population", "Men", "Women","Not Citizen", "Divorced Men", "Divorced Women", "Median Household Income", "Per Capita Income"]
  for col in df.columns[1:-3]:
    if col in integer_cols:
      df[col] = df[col].astype(int)
    else:
      df[col] = df[col].astype(float)

def combine_divorce_cols(df):
  # Calculating the percentage of divorcees, storing the result in the "Divorced Men" column and renaming it accordingly
  divorcee_per = (df["Divorced Men"] + df["Divorced Women"]) * 100 / df["Total Population"]
  divorcee_per = round(divorcee_per, 2)
  df.rename({"Divorced Men": "Divorcee"}, axis=1, inplace=True)
  df["Divorcee"] = divorcee_per
  df.drop(["Divorced Women"], axis=1, inplace=True)

def combine_fips(df):
  df["FIPS"] = df["State"] + df["County"] + df["Tract"]

def sort_by_fips(df):
  df.sort_values(by=["FIPS"], inplace=True)
  df.reset_index(drop=True, inplace=True)

def preprocess_census_df(df):
  convert_columns_dtypes(df)
  combine_divorce_cols(df)
  combine_fips(df)
  sort_by_fips(df)

In [ ]:
preprocess_census_df(ny_df)
ny_df.head()

Name  Total Population   Men  Women  \
0     Census Tract 1, Albany County, New York              2029   890   1139   
1  Census Tract 2.01, Albany County, New York              3263  1427   1836   
2  Census Tract 2.02, Albany County, New York              2153  1059   1094   
3  Census Tract 3.01, Albany County, New York              3016  1347   1669   
4  Census Tract 3.02, Albany County, New York              2931  1287   1644   

   Median Age  White  Black  Native American  Asian  Latin American  \
0        31.9   23.7   56.2              1.4    4.9            10.7   
1        31.2   19.7   68.3              0.0    2.7             4.7   
2        27.7    4.0   91.4              0.0    0.0             3.0   
3        35.6   28.1   37.0              0.0   13.0            32.4   
4        36.2   66.1   16.0              0.0    8.8             2.2   

   Not Citizen  Divorcee  High School Graduate  Bachelor's Degree  \
0          115      7.24                  31.8               11.6   
1          120      3.80                  21.1               23.5   
2           34     11.57                  39.6                6.8   
3          191     13.26                  31.0                2.4   
4          265     13.03                  27.0               14.1   

   Graduate Degree  Median Household Income  Per Capita Income  \
0              7.5                    45268              21252   
1             14.1                    37500              29740   
2              2.5                    25290              17334   
3              6.2                    36964              18922   
4             23.2                    42266              33695   

   Household Income 200k+  Unemployment  \
0                     1.0          10.8   
1                     4.3           3.2   
2                     2.3           9.3   
3                     1.1           6.4   
4                     4.0           3.3   

   Occupation: Management, Business, Science, Art  Occupation: Service  \
0                                            26.9                 36.1   
1                                            46.5                 27.4   
2                                            35.4                 39.0   
3                                            26.1                 40.1   
4                                            54.6                  8.4   

   Occupation: Sales and Office  \
0                          19.2   
1                          15.6   
2                           9.6   
3                          27.2   
4                          26.2   

   Occupation: Natural Resources, Construction, Maintenance  \
0                                                6.0          
1                                                4.4          
2                                                4.0          
3                                                1.0          
4                                                7.1          

   Occupation: Production, Transportation, Material Moving  Private Sector  \
0                                               11.8                  72.2   
1                                                6.1                  64.4   
2                                               12.1                  87.3   
3                                                5.6                  73.6   
4                                                3.8                  63.5   

   Public Sector  Self-Employed  No Health Insurance Coverage  Poverty  \
0           25.3            2.5                           8.8     24.2   
1           26.1            9.5                           3.7     26.9   
2           12.7            0.0                           3.2     34.5   
3           23.2            3.1                           2.9     37.7   
4           34.6            1.9                           1.9     18.7   

   Commuting: Driving Alone  Commuting: Carpool  \
0                      21.0                 8.9   
1          

In [ ]:
ny_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5411 entries, 0 to 5410
Data columns (total 39 columns):
 #   Column                                                    Non-Null Count  Dtype  
---  ------                                                    --------------  -----  
 0   Name                                                      5411 non-null   object 
 1   Total Population                                          5411 non-null   int64  
 2   Men                                                       5411 non-null   int64  
 3   Women                                                     5411 non-null   int64  
 4   Median Age                                                5411 non-null   float64
 5   White                                                     5411 non-null   float64
 6   Black                                                     5411 non-null   float64
 7   Native American                                           5411 non-null   float64
 8   Asian             

All data types were successfully converted and there are no null values, except for 107 in the new divorcee column.

In [ ]:
numeric_cols = ny_df.select_dtypes("number").columns
print("There are " + (ny_df["Total Population"] == 0).sum().astype(str) + " tracts with 0 population.\n")
(ny_df[numeric_cols] < 0).sum()

There are 107 tracts with 0 population.



Total Population                                              0
Men                                                           0
Women                                                         0
Median Age                                                  116
White                                                       107
Black                                                       107
Native American                                             107
Asian                                                       107
Latin American                                              107
Not Citizen                                                   0
Divorcee                                                      0
High School Graduate                                        111
Bachelor's Degree                                           111
Graduate Degree                                             111
Median Household Income                                     202
Per Capita Income                       

In [ ]:
ny_df[ny_df["Total Population"] == 0].head()

Name  Total Population  Men  Women  \
107  Census Tract 19.03, Bronx County, New York                 0    0      0   
108  Census Tract 19.04, Bronx County, New York                 0    0      0   
112     Census Tract 24, Bronx County, New York                 0    0      0   
142  Census Tract 63.02, Bronx County, New York                 0    0      0   
170    Census Tract 110, Bronx County, New York                 0    0      0   

      Median Age        White        Black  Native American        Asian  \
107 -666666666.0 -666666666.0 -666666666.0     -666666666.0 -666666666.0   
108 -666666666.0 -666666666.0 -666666666.0     -666666666.0 -666666666.0   
112 -666666666.0 -666666666.0 -666666666.0     -666666666.0 -666666666.0   
142 -666666666.0 -666666666.0 -666666666.0     -666666666.0 -666666666.0   
170 -666666666.0 -666666666.0 -666666666.0     -666666666.0 -666666666.0   

     Latin American  Not Citizen  Divorcee  High School Graduate  \
107    -666666666.0            0       NaN          -666666666.0   
108    -666666666.0            0       NaN          -666666666.0   
112    -666666666.0            0       NaN          -666666666.0   
142    -666666666.0            0       NaN          -666666666.0   
170    -666666666.0            0       NaN          -666666666.0   

     Bachelor's Degree  Graduate Degree  Median Household Income  \
107       -666666666.0     -666666666.0               -666666666   
108       -666666666.0     -666666666.0               -666666666   
112       -666666666.0     -666666666.0               -666666666   
142       -666666666.0     -666666666.0               -666666666   
170       -666666666.0     -666666666.0               -666666666   

     Per Capita Income  Household Income 200k+  Unemployment  \
107         -666666666            -666666666.0  -666666666.0   
108         -666666666            -666666666.0  -666666666.0   
112         -666666666            -666666666.0  -666666666.0   
142         -666666666            -666666666.0  -666666666.0   
170         -666666666            -666666666.0  -666666666.0   

     Occupation: Management, Business, Science, Art  Occupation: Service  \
107                                    -666666666.0         -666666666.0   
108                                    -666666666.0         -666666666.0   
112                                    -666666666.0         -666666666.0   
142                                    -666666666.0         -666666666.0   
170                                    -666666666.0         -666666666.0   

     Occupation: Sales and Office  \
107                  -666666666.0   
108                  -666666666.0   
112                  -666666666.0   
142                  -666666666.0   
170                  -666666666.0   

     Occupation: Natural Resources, Construction, Maintenance  \
107                                       -666666666.0          
108                                       -666666666.0          
112                                       -666666666.0          
142                                       -666666666.0          
170                                       -666666666.0          

     Occupation: Production, Transportation, Material Moving  Private Sector  \
107                                       -666666666.0          -666666666.0   
108                                       -666666666.0          -666666666.0   
112                                       -666666666.0          -666666666.0   
142                                       -666666666.0          -666666666.0   
170                                       -666666666.0          -666666666.0   

     Public Sector  Self-Employed  No Health Insurance Coverage      Poverty  \
107   -666666666.0   -666666666.0                  -666666666.0 -666666666.0   
108   -666666666.0   -666666666.0                  -666666666.0 -666666666.0   
112   -666666666.0   -666666666.0                  -666666666.0 -666666666.0   
142   -666666666.0   -66

Note that there are some negative values in the dataframe. Taking a deeper look at the tracts with no population (these are places such as parks and airports) we can see that the value -666666666 was chosen as a replacement for NaN when there was no data. That's why we have null values in the new divorcee column.

For now, we will convert those values to 0. We may choose to exclude these tracts from our analysis later on.

Three exceptions are median household income, per capita income and mean commute time - there, we will replace NaN values with the median where the population is greater than 0.

In [ ]:
def clean_missing_values(df):
  df.fillna(0, inplace = True)

  # Calculating the median according to the correct values
  per_capita_median = df[(df["Total Population"] > 0) & (df["Per Capita Income"] > 0)]["Per Capita Income"].median()
  household_median = df[(df["Total Population"] > 0) & (df["Median Household Income"] > 0)]["Median Household Income"].median()
  commute_median = df[(df["Total Population"] > 0) & (df["Mean Commute Time (Minutes)"] > 0)]["Mean Commute Time (Minutes)"].median()

  df.loc[(df["Total Population"] > 0) & (df["Per Capita Income"] < 0), "Per Capita Income"] = per_capita_median
  df.loc[(df["Total Population"] > 0) & (df["Median Household Income"] < 0), "Median Household Income"] = household_median
  df.loc[(df["Total Population"] > 0) & (df["Mean Commute Time (Minutes)"] < 0), "Mean Commute Time (Minutes)"] = commute_median
  
  # Changing all the remaining negative values to 0
  num = df._get_numeric_data()
  num[num < 0] = 0

In [ ]:
clean_missing_values(ny_df)
(ny_df[numeric_cols] < 0).sum()

Total Population                                            0
Men                                                         0
Women                                                       0
Median Age                                                  0
White                                                       0
Black                                                       0
Native American                                             0
Asian                                                       0
Latin American                                              0
Not Citizen                                                 0
Divorcee                                                    0
High School Graduate                                        0
Bachelor's Degree                                           0
Graduate Degree                                             0
Median Household Income                                     0
Per Capita Income                                           0
Househol

We have successfully cleaned the census data!

Now, we would like to add the 2020 election results data to our dataset. Harvard University has published the [US Voting by Census Block Groups Dataset](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/NKNWBX) which has exactly the data that we need. However, the data is available for smaller geographical units. As such, in order to get voting data for a census tract we will have to sum the results of the census block groups that constitute it.

In [ ]:
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
path = "Your path/2021blockgroupvoting.csv"
election_df = pd.read_csv(path)
election_df

REP         DEM        LIB        OTH          AREA  \
0        284.224647  408.322732  18.014238  14.811707  7.144464e+07   
1       1215.499906  580.758667  59.983880  26.981497  7.348527e+08   
2       1063.299740  268.872878  41.752343   8.671476  1.978857e+08   
3        233.737769   58.902452   9.528108   2.294799  4.098606e+07   
4        246.541377   91.700706  12.587948   3.776375  5.149964e+07   
...             ...         ...        ...        ...           ...   
233861   125.339970  223.155960  12.703375   8.468917  6.127563e+05   
233862   220.233407  441.311959  24.380961  12.706912  4.293685e+06   
233863   270.483041  376.756600  32.140024  14.214931  1.743156e+06   
233864   235.380599  277.381764  27.098904  10.732766  1.414571e+07   
233865    40.182380   25.906145   2.158845   1.183883  1.400019e+06   

                 GAP  PRECINCTS STATE  BLOCKGROUP_GEOID  
0       2.980232e-08          3    AK       20900013001  
1       1.668930e-06          6    AK       20900019022  
2      -8.940697e-08          3    AK       20900015011  
3       7.450581e-09          4    AK       20900015021  
4      -1.266599e-07          3    AK       20900014011  
...              ...        ...   ...               ...  
233861  0.000000e+00          4    WY      560019636002  
233862 -2.514571e-08          4    WY      560019629002  
233863 -1.164153e-09          3    WY      560019630001  
233864  7.450581e-09          3    WY      560019630002  
233865 -9.313226e-10          2    WY      560019628002  

[233866 rows x 9 columns]

In [ ]:
election_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 233866 entries, 0 to 233865
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   REP               233866 non-null  float64
 1   DEM               233866 non-null  float64
 2   LIB               233866 non-null  float64
 3   OTH               233866 non-null  float64
 4   AREA              233866 non-null  float64
 5   GAP               233866 non-null  float64
 6   PRECINCTS         233866 non-null  int64  
 7   STATE             233866 non-null  object 
 8   BLOCKGROUP_GEOID  233866 non-null  int64  
dtypes: float64(6), int64(2), object(1)
memory usage: 16.1+ MB


In this dataset votes are attributed based on the proportion of the precinct's area that intersects the corresponding block group (zones with different geographical boundaries). Therefore, the number of votes is not an integer.

We are interested in the share of vote for the Democrat and the Republican Parties in each tract. Other columns will be dropped. As the results are for blocks - which are smaller geographical units that compose a tract - we will omit the last digit of the 12-digit FIPS code that corresponds to the block. Then, we will merge the results of the rows that share the same 11-digit code that represents the census tract.

In [ ]:
def get_state_election_results(df, chosen_state, add_leading_zero=False):
  """
  chosen_state (str): Two letters representing the state (for example: NY, CA, TX)
  """
  state_df = df[df["STATE"] == chosen_state].copy()
  state_df["BLOCKGROUP_GEOID"] = state_df["BLOCKGROUP_GEOID"] // 10

  state_df = state_df.groupby("BLOCKGROUP_GEOID").sum().reset_index()

  total_votes = state_df["LIB"] + state_df["OTH"] + state_df["REP"] + state_df["DEM"]
  state_df.drop(["LIB", "OTH", "AREA", "GAP", "PRECINCTS"], axis=1, inplace=True)

  percent_dem = state_df["DEM"] * 100 / total_votes
  percent_rep = state_df["REP"] * 100 / total_votes

  state_df["DEM"] = round(percent_dem, 2)
  state_df["REP"] = round(percent_rep, 2)

  state_df.sort_values(by=["BLOCKGROUP_GEOID"], inplace=True)
  state_df["BLOCKGROUP_GEOID"] = state_df["BLOCKGROUP_GEOID"].astype(str)
  state_df.rename({"BLOCKGROUP_GEOID": "FIPS"}, axis=1, inplace=True)

  if add_leading_zero:
    state_df['FIPS'] = '0' + state_df['FIPS']

  return state_df

In [ ]:
ny_election_df = get_state_election_results(election_df, "NY")
ny_election_df

FIPS    REP    DEM
0     36001000100  13.64  85.49
1     36001000201   7.79  90.36
2     36001000202   8.42  90.58
3     36001000301  12.80  84.87
4     36001000302  24.31  74.12
...           ...    ...    ...
5406  36123150301  64.94  32.86
5407  36123150302  59.76  37.56
5408  36123150400  51.93  44.88
5409  36123150501  51.91  45.54
5410  36123150502  58.72  38.90

[5411 rows x 3 columns]

In [ ]:
ny_election_df.isnull().sum()

FIPS     0
REP     12
DEM     12
dtype: int64

Now we have some null values. These are probably blocks that recorded no votes. Let's check.

In [ ]:
ny_election_df[ny_election_df.isnull().any(axis=1)]

FIPS  REP  DEM
101   36005000100  NaN  NaN
567   36011990200  NaN  NaN
604   36013990000  NaN  NaN
2013  36047990100  NaN  NaN
2568  36059990100  NaN  NaN
2569  36059990200  NaN  NaN
2570  36059990301  NaN  NaN
2571  36059990302  NaN  NaN
2948  36063990000  NaN  NaN
3356  36075990000  NaN  NaN
4127  36081990100  NaN  NaN
4301  36085990100  NaN  NaN

In [ ]:
election_df[election_df['BLOCKGROUP_GEOID'].astype(str).str.contains("36005000100")]

REP  DEM  LIB  OTH          AREA           GAP  PRECINCTS STATE  \
155575  0.0  0.0  0.0  0.0  1.800647e+06  1.773395e+06          2    NY   
155583  0.0  0.0  0.0  0.0  2.917713e+06  9.222974e+03          3    NY   

        BLOCKGROUP_GEOID  
155575      360050001000  
155583      360050001001

Looking at one of these FIPS codes in the general voting dataframe suggests that our assumption is correct. We will convert these values to 0.

In [ ]:
ny_election_df.fillna(0, inplace = True)

It only remains to merge the dataframes and save the resulting one as a csv file.

In [ ]:
new_york_df = pd.merge(ny_df, ny_election_df, on="FIPS")
new_york_df.rename({"REP": "Republican", "DEM": "Democrat"}, axis=1, inplace=True)

In [ ]:
new_york_df[new_york_df["Total Population"] == 0].head()

Name  Total Population  Men  Women  \
107  Census Tract 19.03, Bronx County, New York                 0    0      0   
108  Census Tract 19.04, Bronx County, New York                 0    0      0   
112     Census Tract 24, Bronx County, New York                 0    0      0   
142  Census Tract 63.02, Bronx County, New York                 0    0      0   
170    Census Tract 110, Bronx County, New York                 0    0      0   

     Median Age  White  Black  Native American  Asian  Latin American  \
107         0.0    0.0    0.0              0.0    0.0             0.0   
108         0.0    0.0    0.0              0.0    0.0             0.0   
112         0.0    0.0    0.0              0.0    0.0             0.0   
142         0.0    0.0    0.0              0.0    0.0             0.0   
170         0.0    0.0    0.0              0.0    0.0             0.0   

     Not Citizen  Divorcee  High School Graduate  Bachelor's Degree  \
107            0       0.0                   0.0                0.0   
108            0       0.0                   0.0                0.0   
112            0       0.0                   0.0                0.0   
142            0       0.0                   0.0                0.0   
170            0       0.0                   0.0                0.0   

     Graduate Degree  Median Household Income  Per Capita Income  \
107              0.0                        0                  0   
108              0.0                        0                  0   
112              0.0                        0                  0   
142              0.0                        0                  0   
170              0.0                        0                  0   

     Household Income 200k+  Unemployment  \
107                     0.0           0.0   
108                     0.0           0.0   
112                     0.0           0.0   
142                     0.0           0.0   
170                     0.0           0.0   

     Occupation: Management, Business, Science, Art  Occupation: Service  \
107                                             0.0                  0.0   
108                                             0.0                  0.0   
112                                             0.0                  0.0   
142                                             0.0                  0.0   
170                                             0.0                  0.0   

     Occupation: Sales and Office  \
107                           0.0   
108                           0.0   
112                           0.0   
142                           0.0   
170                           0.0   

     Occupation: Natural Resources, Construction, Maintenance  \
107                                                0.0          
108                                                0.0          
112                                                0.0          
142                                                0.0          
170                                                0.0          

     Occupation: Production, Transportation, Material Moving  Private Sector  \
107                                                0.0                   0.0   
108                                                0.0                   0.0   
112                                                0.0                   0.0   
142                                                0.0                   0.0   
170                                                0.0                   0.0   

     Public Sector  Self-Employed  No Health Insurance Coverage  Poverty  \
107            0.0            0.0                           0.0      0.0   
108            0.0            0.0                           0.0      0.0   
112            0.0            0.0                           0.0      0.0   
142            0.0            0.0                           0.0      0.0   
170            0.0            0.0                           0.0      0.0   

 

Some tracts have no population yet do have voting data. The reason for it is probably the way the election dataset was created and therefore we will keep these results. We can ignore them later.

In [ ]:
new_york_df.to_csv('/content/drive/MyDrive/Colab Notebooks/US Census/Data/Preprocessed Data/new_york_data.csv', index=False)

We will now repeat the same process to retrieve data for California. There is just a small extra step: as the FIPS column in the election results dataframe initially contains integers, FIPS codes start with 6 rather than 06 and we have to add that leading zero.

In [ ]:
def census_data_pipeline(API_KEY, get_vars_dict, states):
  df = get_census_data(API_KEY, get_vars_dict, states)
  preprocess_census_df(df)
  clean_missing_values(df)
  return df

def election_data_pipeline(df, chosen_state, add_leading_zero=False):
  """
  df: A dataframe containing census data
  chosen_state (str): Two letters representing the state (for example: NY, CA, TX)
  """
  state_df = get_state_election_results(election_df, chosen_state, add_leading_zero)
  state_df.fillna(0, inplace = True)
  merged_df = pd.merge(df, state_df, on="FIPS")
  merged_df.rename({"REP": "Republican", "DEM": "Democrat"}, axis=1, inplace=True)
  return merged_df

In [ ]:
states = ["06"]
cali_df = census_data_pipeline(API_KEY, get_vars_dict, states)
california_df = election_data_pipeline(cali_df, 'CA', add_leading_zero=True)
california_df.to_csv('Your path/california_data.csv', index=False)

In [ ]:
states = ["48"]
tx_df = census_data_pipeline(API_KEY, get_vars_dict, states)
texas_df = election_data_pipeline(tx_df, 'TX')
texas_df.to_csv('Your path/texas_data.csv', index=False)

In [ ]:
states = ["12"]
fl_df = census_data_pipeline(API_KEY, get_vars_dict, states)
florida_df = election_data_pipeline(fl_df, 'FL')
florida_df.to_csv('Your path/florida_data.csv', index=False)